### ISSUE TITLE
Upgrading to 3.0.1: Text search failed. fatal: not a git repository

### ISSUE THREAD

**Issue by KattMingMing** [sourcegraph/sourcegraph#2243]
- **Sourcegraph version:** 3.0.1
- External postgres via AWS RDS

After upgrading from 3.0.0-beta to 3.0.1 text search fails with the following error (redacted company info):

```
error occurred: * text search failed: failed to search github.com/REDACTED_ORG/REDACTED_REPO@: git command [git rev-parse HEAD] failed (stderr: "fatal: not a git repository (or any parent up to mount point /var/opt) Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set)."): exit status 128 
```

**Comment by keegancsmith** [comment]
@KattMingMing did this issue keep occurring? Sorry for missing this issue when you originally filed it.

**Comment by dadlerj** [comment]
Not sure who the original issue was in response to, but it seems like another org is running into this issue:
* https://app.hubspot.com/contacts/2762526/company/1768250576

They are on version 3.3.1, I don't believe they have an external database.

Deleting the repo helps, but it "creeps up from time to time" — I'm clarifying whether that means the same repo is reindexed or if it affects other repos.

I asked the user to clone one of the offending repos locally and to try to run git commands to see if the repo really does have malformed (or nonexistent) git data.

**Comment by tsenart** [comment]
@keegancsmith: Were there any changes between 3.3.1 and 3.6 that could alleviate this issue? My spidey sense tells me this has something to do with non-atomic file-system operations.

**Comment by keegancsmith** [comment]
I'm not sure, there was a brief amount of time we didn't use atomic git operations for cleanup (but I don't know if that went out in a release, or was just some brief time in master). Otherwise in general we only use atomic operations (clone to tmp dir, rename into target location; rename into tmp dir, rm -rf).

To be honest it would be great if the customer running into this could give us exact logs privately of this happening and more information. Even better would be if they upgraded to 3.6 first :)

**Comment by mrnugget** [comment]
(I wrote the following on Slack, but want to document it here too:)

The error message sounds to me like we're trying to run git commands in directories without .git folders. You can reproduce the exact same error code by running git in a non-git-dir:

```
~/tmp λ git status
fatal: not a git repository (or any of the parent directories): .git
~/tmp λ echo $?
128
```


Now the question is: do we create directories in the gitserver data directory that look like git clones, but aren't git clones? If I'm not mistaken, the answer might be "yes":

https://github.com/sourcegraph/sourcegraph/blob/e1b142b7a7ee850c9a9ae06b4dc8679680043172/cmd/gitserver/server/server.go#L833-L838

Here, when we're done cloning a repo, we first do a `os.MkdirAll`, which creates the folder that looks like a repo and only then do we copy the actual clone to this folder.

In between line 833 and 836 we have a folder in our repos folder that does not have a .git folder.

---

I added some debug print and sleep statements to the code to confirm this:

```
gitserver | INFO cloning repo, repo: github.com/mrnugget/fzz, tmp: /Users/thorstenball/.sourcegraph/repos/.tmp/clone-065458346/.git, dst: /Users/thorstenball/.sourcegraph/repos/github.com/mrnugget/fzz/.git

gitserver | dstPath=/Users/thorstenball/.sourcegraph/repos/github.com/mrnugget/fzz/.git
gitserver | tmpPath=/Users/thorstenball/.sourcegraph/repos/.tmp/clone-065458346/.git

gitserver | os.MkdirAll(/Users/thorstenball/.sourcegraph/repos/github.com/mrnugget/fzz, os.ModePerm)

gitserver | going to sleep...
gitserver | back from sleep.

gitserver | renameAndSync(/Users/thorstenball/.sourcegraph/repos/.tmp/clone-065458346/.git, /Users/thorstenball/.sourcegraph/repos/github.com/mrnugget/fzz/.git)

gitserver | INFO repo cloned, repo: github.com/mrnugget/fzz

```

So, yes, between the `os.MkdirAll` and the `renameAndSync` the directory `/Users/thorstenball/.sourcegraph/repos/github.com/mrnugget/fzz` exists, but it doesn't have a `.git` folder.

Now, the usual search in this protected says "clone in progress"... But what if other operations do not check whether a clone is in progress?

**Comment by unknwon** [comment]
Just curious... it feels mysterious why the error is not `os.PathError` but Git command error? If the `dstPath=/Users/thorstenball/.sourcegraph/repos/github.com/mrnugget/fzz/.git` certainly not exists before `renameAndSync` completes so then which directory the Git command was executed?

**Comment by keegancsmith** [comment]
However, we won't run a command against a repo unless it exists:

https://github.com/sourcegraph/sourcegraph/blob/e1b142b7a7ee850c9a9ae06b4dc8679680043172/cmd/gitserver/server/server.go#L615

https://github.com/sourcegraph/sourcegraph/blob/e1b142b7a7ee850c9a9ae06b4dc8679680043172//cmd/gitserver/server/serverutil.go#L78-L86

**Comment by mrnugget** [comment]
> However, we won't run a command against a repo unless it exists:

That's what I discovered after posting the above comment. Every access to the directory is "locked" as long as it's being cloned.

The only possibilities of bypassing the `repoCloned` check and the in-memory `*RepositoryLocker` that I can see are:
1. `gitserver` crashes while cloning. Information in `*RepositoryLocker` gets lost (returns false for repos). `s.locker.Status` returns false for `cloneInProgress` and `repoCloned` also returns false because the two files do not exist yet.
2. The cleanup tasks run when the repository is being cloned. AFAICT they don't use `repoCloned` or `s.Locker.Status`. See `maybeReclone` here: https://github.com/sourcegraph/sourcegraph/blob/91127d71302bd17060ba11ec50a2201a5372a7b7/cmd/gitserver/server/cleanup.go#L78-L107

That being said: we fixed the customer issue by deleting the repositories and re-cloning them. That fixed it. And the weird thing is that the repos _did_ look like git repos: the `.git` dir was there, the `HEAD` file was there, etc. But still git reported that it's not a git directory.
